src/server/scripts/Outland/CoilfangReservoir/TheUnderbog/boss_hungarfen.cpp

```diff
@@ -15,166 +15,150 @@
-
-                Mushroom_Timer = 10000;
-            } else Mushroom_Timer -= diff;
-
-            if (AcidGeyser_Timer <= diff)
+            _scheduler.Schedule(3s, 5s, [this](TaskContext task)
             {
                 if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0))
                     DoCast(target, SPELL_ACID_GEYSER);
-                AcidGeyser_Timer = 10000 + rand32() % 7500;
-            } else AcidGeyser_Timer -= diff;
-
-            DoMeleeAttackIfReady();
+                task.Repeat(10s, 15s);
+            });
         }
-    };
-
-};
+    }
 
-class npc_underbog_mushroom : public CreatureScript
-{
-public:
-    npc_underbog_mushroom() : CreatureScript("npc_underbog_mushroom") { }
+    void EnterEvadeMode(EvadeReason why) override
+    {
+        DoCastSelf(SPELL_DESPAWN_UNDERBOG_MUSHROOMS, true);
+        ScriptedAI::EnterEvadeMode(why);
+    }
 
-    CreatureAI* GetAI(Creature* creature) const override
+    void JustDied(Unit* /*killer*/) override
     {
-        return GetTheUnderbogAI<npc_underbog_mushroomAI>(creature);
+        DoCastSelf(SPELL_DESPAWN_UNDERBOG_MUSHROOMS, true);
     }
 
-    struct npc_underbog_mushroomAI : public ScriptedAI
+    void UpdateAI(uint32 diff) override
     {
-        npc_underbog_mushroomAI(Creature* creature) : ScriptedAI(creature)
-        {
-            Initialize();
-        }
+        if (!UpdateVictim())
+            return;
 
-        void Initialize()
+        _scheduler.Update(diff, [this]
         {
-            Stop = false;
-            Grow_Timer = 0;
-            Shrink_Timer = 20000;
-        }
-
-        bool Stop;
-        uint32 Grow_Timer;
-        uint32 Shrink_Timer;
+            DoMeleeAttackIfReady();
+        });
 
-        void Reset() override
+        if (!HealthAbovePct(20) && !_roared)
         {
-            Initialize();
+            Talk(EMOTE_ROARS);
+            _roared = true;
+            me->SetReactState(REACT_PASSIVE);
 
-            DoCast(me, SPELL_PUTRID_MUSHROOM, true);
-            DoCast(me, SPELL_SPORE_CLOUD, true);
+            _scheduler.Schedule(2s, [this](TaskContext /*task*/)
+            {
+                DoCastSelf(SPELL_FOUL_SPORES);
+                me->SetReactState(REACT_AGGRESSIVE);
+            });
         }
+    }
 
-        void MoveInLineOfSight(Unit* /*who*/) override { }
+private:
+    TaskScheduler _scheduler;
+    bool _roared;
+};
 
-        void AttackStart(Unit* /*who*/) override { }
+struct npc_underbog_mushroom : public ScriptedAI
+{
+    npc_underbog_mushroom(Creature* creature) : ScriptedAI(creature) { }
 
-        void JustEngagedWith(Unit* /*who*/) override { }
+    void InitializeAI() override
+    {
+        _counter = RAND(8, 9, 10);
+        me->SetReactState(REACT_PASSIVE);
+        DoCastSelf(SPELL_SHRINK);
```

Comment:
I'd move spell casting related stuff to JustAppeared

Reply:
Isn't it slower?

Reply:
Nope, it makes them shrink after first update tick, it's something you will not see on retail
It's not just visual aura or something, it makes creature change scale and if it will be not called instantly, creature will first spawn with normal scale and then shrink